File: Baekjoon_Online_Judge/NO_02784/2784.cpp
#include <iostream>
using namespace std;
/*
 -------------------------------------- 
 | Created by happiness96             | 
 | Year 2020                          | 
 | Month 10                           | 
 | Day 31                             | 
 | 2784 가로 세로 퍼즐                 | 
 -------------------------------------- 
*/

int main(){
    string words[6];
    string result = "ZZZZZZZZZZZZ";

    for(int i = 0; i < 6; i ++){
        cin >> words[i];
    }
    
    for(int i = 0; i < 6; i ++){
        for(int j = 0; j < 6; j ++){
            if(i == j) continue;

            for(int k = 0; k < 6; k ++){
                if(k == i or k == j) continue;
                string tmp[3] = {words[i], words[j], words[k]};

                int chk[6] = {0, };
                int flag = 0;
                chk[i] = 1;
                chk[j] = 1;
                chk[k] = 1;
                
                for(int col = 0; col < 3; col ++){
                    string save = "";

                    for(int row = 0; row < 3; row ++) save += tmp[row][col];

                    for(int chk_ind = 0; chk_ind < 6; chk_ind ++){
                        if(chk[chk_ind] == 0 and save == words[chk_ind]){
                            chk[chk_ind] = 1;
                            flag += 1;
                            break;
                        }
                    }
                }

                if(flag == 3){
                    if(result > words[i] + words[j] + words[k]) result = words[i] + words[j] + words[k];
                }

            }
        }
    }

    if(result == "ZZZZZZZZZZZZ") cout << 0;
    else{
        for(int i = 0; i < 9; i += 3){
            cout << result[i] << result[i + 1] << result[i + 2] << "\n";
        }
    }
    return 0;
}

File: Baekjoon_Online_Judge/NO_02784/2784.py
# -*- encoding: utf-8 -*-
import sys
r_input = sys.stdin.readline

# -------------------------------------- #
# | Created by happiness96             | #
# | Year 2020                          | #
# | Month 10                           | #
# | Day 31                             | #
# | 2784 가로 세로 퍼즐                | #
# -------------------------------------- #


def run():
    words = [r_input().rstrip() for _ in range(6)]

    result = []

    for ind1, word1 in enumerate(words):
        for ind2, word2 in enumerate(words):
            if ind1 != ind2:
                for ind3, word3 in enumerate(words):
                    if ind1 != ind3 and ind2 != ind3:
                        save = [word1, word2, word3]

                        chk = list(range(6))
                        chk.remove(ind1)
                        chk.remove(ind2)
                        chk.remove(ind3)

                        for col in range(3):
                            string = ''

                            for row in range(3):
                                string += save[row][col]
                            
                            for k in chk:
                                if string == words[k]:
                                    chk.remove(k)
                                    break
                        
                        if not chk:
                            result.append(word1 + word2 + word3)
    
    result.sort()
    
    if not result:
        print(0)
        sys.exit()

    for res in range(0, 9, 3):
        print(result[0][res: res + 3])


if __name__ == "__main__":
    run()
